sorc/global_cycle.fd/cycle.f90

```diff
@@ -688,25 +724,65 @@ SUBROUTINE SFCDRV(LUGB, IDIM,JDIM,LENSFC,LSOIL,DELTSFC,  &
         SLC_BCK = SLCFCS
 
         ! below updates [STC/SMC/STC]FCS to hold the analysis
-        CALL ADD_INCREMENT_SOIL(RLA,RLO,STCFCS,SMCFCS,SLCFCS,STC_UPDATED, SLC_UPDATED, &
+        ! ONLY isolate soil temperature and soil moisture ingest/adjust routines in
+        ! the jedi-related portion. In gsi-related part, it is *not* updated yet.
+        ! The gsi-related part is still using the old assimilation sequence
+        ! (i.e., soil moisture incr. ingest based on background state, not the
+        ! analysis state as edited in the new assimilation sequence)
+        ! The new assimilation/adjustment sequence in jedi is: 
+        ! a) ingest soil temp incr. 
+        ! b) adjust soil ice/liquid moisture as needed
+        ! c) ingest soil moisure incr. based on newly-adjusted soil state due to b)
+        ! d) ingest soil moisture incr.
+        ! e) adjust/bound soil moisture
+        IF (DO_SOI_INC_GSI) THEN
+          CALL ADD_GSI_INCREMENT_SOIL(RLA,RLO,STCFCS,SMCFCS,SLCFCS,STC_UPDATED, SLC_UPDATED, &
+                STC_INC_TMP, SLC_INC_TMP, &
                 LANDINC_MASK,LANDINC_MASK_FG,LENSFC,LSOIL,IDIM,JDIM,LSM,MYRANK)
 
         !--------------------------------------------------------------------------------
         ! make any necessary adjustments to dependent variables
         !--------------------------------------------------------------------------------
 
+          CALL APPLY_LAND_DA_ADJUSTMENTS_SOIL_TEMP(LSM, ISOT, IVEGSRC, LENSFC,&
+                                               LSOIL, SOTFCS, LANDINC_MASK_FG,&
+                                               STC_BCK, STCFCS, SMCFCS, SLCFCS,&
+                                               STC_UPDATED)
+          CALL APPLY_LAND_DA_ADJUSTMENTS_SOIL_MOISTURE(LSM, LENSFC, &
+                                              LSOIL, SMCFCS, SLCFCS,&
+                                              SLC_UPDATED, ZSOIL)
+
+        ! if read in GSI-based increments, save interpolated fields onto cubed sphere tiles
+          CALL WRITE_DATA(LENSFC,IDIM,JDIM,LSOIL,DO_NSST,.true.,NSST, &
+                   SLC_INC_TMP=SLC_INC_TMP,STC_INC_TMP=STC_INC_TMP)
+
+        ELSE IF (DO_SOI_INC_JEDI) THEN
+          CALL ADD_JEDI_INCREMENT_SOIL_TEMP(STCINC, STCFCS, &
```

Comment:
Is it possible (or I really mean "easy") to generalize the program that adds the increments to take either input, then interpolate the GSI increment if necessary?